docs/4.4/api-reference.md

```diff
@@ -111,62 +108,317 @@ func main() {
+type RoleV3 struct {
+  // Resource Fields are fields that all Teleport resources have - see above
+  Resource Fields
+  // Spec is the role specification
+  Spec RoleSpecV3 struct {
+    // Options is for OpenSSH options like agent forwarding.
+    Options RoleOptions
+    // Allow is the set of conditions evaluated to grant access.
+    Allow RoleConditions
+    // Deny is the set of conditions evaluated to deny access. Deny takes priority over allow.
+    Deny RoleConditions
+  }
+}
+```
+
+### Role Options
+
+The `RoleOptions` struct defines what OpenSSH actions a user is allowed to use.
+
+```go
+// RoleOptions is a set of role options
+type RoleOptions struct {
+  // ForwardAgent is SSH agent forwarding. default true.
+  ForwardAgent Bool
+  // MaxSessionTTL defines how long a SSH session can last for.
+  MaxSessionTTL Duration
+  // PortForwarding defines if the certificate will have "permit-port-forwarding"
+  // in the certificate. PortForwarding is true if not set.
+  PortForwarding *BoolOption // struct { Value bool } - Nullable boolean
+  CertificateFormat string
+  // ClientIdleTimeout sets disconnect clients on idle timeout behavior,
+  // if set to 0 means do not disconnect, otherwise is set to the idle duration.
+  ClientIdleTimeout Duration
+  // DisconnectExpiredCert sets disconnect clients on expired certificates.
+  DisconnectExpiredCert Bool
+  // BPF defines what events to record for the BPF-based session recorder.
+  BPF []string
+  // PermitX11Forwarding authorizes use of X11 forwarding.
+  PermitX11Forwarding Bool
+  // MaxConnections defines the maximum number of
+  // concurrent connections a user may hold.
+  MaxConnections int64
+  // MaxSessions defines the maximum number of
+  // concurrent sessions per connection.
+  MaxSessions int64
+}
+```
+
+### Role Conditions
+
+The `RoleConditions` struct allows for precise permission combinations between a role's `Allow` and `Deny` fields.
+
+`Deny` rules take priority over `Allow` rules. If a user has multiple roles, their role permissions will be evaluated together, which can potentially lead to conflicting rules. This can result in a user being denied permissions that another role of theirs allows.
+
+However, this also allows you to take a modular approach to defining roles. Splitting roles up into logical parts will allow you to manage roles for many developers more effectively. It may also be effective to keep your deny and allow rules in separate roles so that conflicting roles are obvious.
+
+```go
+// RoleConditions is a set of conditions that must all match to be allowed or denied access.
+type RoleConditions struct {
+  // Logins is a list of *nix system logins,  e.g. "root".
+  Logins []string
+  // Namespaces is a list of namespaces (used to partition a cluster).
+  Namespaces []string
+  // NodeLabels is a map of node labels (used to dynamically grant access to nodes).
+  NodeLabels Labels // essentially map[string][]string
```

Comment:
I notice that with some of these fields, we're giving the name of a struct and what it actually consists of (like with `NodeLabels Labels` here where we say `// essentially map[string][]string`) but on some of these fields we're explicitly saying `struct { Roles []string }`, like with `Request *AccessRequestConditions` below.

Is this because `Request` actually accepts a generic struct whereas `NodeLabels` has a requirement of being that named struct, or is this an inconsistency?

Reply:
Labels are just a custom type for map[string]Strings, which is essentially map[string][]string and doesn't need much explanation imo. But I did handle other structs in a somewhat inconsistent way, because some needed more explanation than others. I'll mess around with some different options to handle this issue.